File: packages/guards/src/token-limit.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

interface TokenLimitOptions {
  action: 'block' | 'warn';
  maxTokens: number;
  estimationMethod?: 'chars' | 'words' | 'tiktoken-approx';
}

function estimateTokens(text: string, method: string): number {
  switch (method) {
    case 'chars':
      return Math.ceil(text.length / 4);
    case 'words':
      return Math.ceil(text.split(/\s+/).filter((w) => w.length > 0).length * 1.3);
    case 'tiktoken-approx':
    default: {
      const words = text.split(/\s+/).filter((w) => w.length > 0);
      let tokens = 0;
      for (const w of words) {
        if (w.length <= 4) tokens += 1;
        else if (w.length <= 8) tokens += 2;
        else tokens += Math.ceil(w.length / 4);
      }
      return tokens;
    }
  }
}

export function tokenLimit(options: TokenLimitOptions): Guard {
  const method = options.estimationMethod ?? 'tiktoken-approx';

  return {
    name: 'token-limit',
    version: '0.1.0',
    description: 'Token count estimation and limit guard',
    category: 'format',
    supportedStages: ['input', 'output'],
    async check(
      text: string,
      _ctx: GuardContext,
    ): Promise<GuardResult> {
      const start = performance.now();
      const estimated = estimateTokens(text, method);
      const triggered = estimated > options.maxTokens;

      return {
        guardName: 'token-limit',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: {
          estimatedTokens: estimated,
          maxTokens: options.maxTokens,
          method,
        },
      };
    },
  };
}

File: python/open_guardrail/guards/token_limit.py
"""Enforces approximate token count limits."""
from __future__ import annotations

import time
from typing import Optional

from open_guardrail.core import GuardResult


class _TokenLimit:
    def __init__(
        self, *, action: str = "block", max_tokens: int = 4096, chars_per_token: float = 4.0,
    ) -> None:
        self.name = "token-limit"
        self.action = action
        self.max_tokens = max_tokens
        self.chars_per_token = chars_per_token

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        estimated = int(len(text) / self.chars_per_token + 0.5)
        triggered = estimated > self.max_tokens
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="token-limit",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=f"Token limit exceeded: ~{estimated} > {self.max_tokens}" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"estimated_tokens": estimated, "max_tokens": self.max_tokens} if triggered else None,
        )


def token_limit(
    *, action: str = "block", max_tokens: int = 4096, chars_per_token: float = 4.0,
) -> _TokenLimit:
    return _TokenLimit(action=action, max_tokens=max_tokens, chars_per_token=chars_per_token)
